src/ribasim_nl/ribasim_nl/network.py

```diff
@@ -135,58 +202,480 @@ def add_link(
+                nodes_select["distance"] = nodes_select.geometry.apply(
+                    lambda x: geometry.project(x)
                 )
                 nodes_select.sort_values("distance", inplace=True)
 
                 # More than one node. We select start_node and point-geometry
-                node_from = nodes_select.index[0]
-                point_from = nodes_select.loc[node_from].geometry
+                link_def["node_from"] = nodes_select.index[0]
+                link_def["point_from"] = nodes_select.loc[
+                    link_def["node_from"]
+                ].geometry
 
                 # More than two nodes. Line should be split into parts. We create one extra edge for every extra node
                 if len(nodes_select) > 2:
                     for node in nodes_select[1:-1].itertuples():
-                        node_to = node.Index
-                        point_to = nodes_select.loc[node_to].geometry
+                        link_def["node_to"] = node.Index
+                        link_def["point_to"] = nodes_select.loc[
+                            link_def["node_to"]
+                        ].geometry
                         edge_geometry, geometry = split(
-                            snap(geometry, point_to, self.snap_tolerance), point_to
+                            snap(geometry, link_def["point_to"], self.snap_tolerance),
+                            link_def["point_to"],
                         ).geoms
-                        add_link(
-                            node_from,
-                            point_from,
-                            node_to,
-                            point_to,
-                            edge_geometry,
-                            links_data,
-                        )
-                        node_from = node_to
-                        point_from = point_to
+                        link_def["geometry"] = edge_geometry
+                        self.add_link(**link_def)
+                        link_def["node_from"] = link_def["node_to"]
+                        link_def["point_from"] = link_def["point_to"]
 
                 # More than one node. We finish the (last) edge
-                node_to = nodes_select.index[-1]
-                point_to = nodes_select.loc[node_to].geometry
-                add_link(
-                    node_from,
-                    point_from,
-                    node_to,
-                    point_to,
-                    geometry,
-                    links_data,
-                )
+                link_def["node_to"] = nodes_select.index[-1]
+                link_def["point_to"] = nodes_select.loc[link_def["node_to"]].geometry
+                link_def["geometry"] = geometry
+                self.add_link(**link_def)
+
+            # Set all node-types
+            self.set_node_types()
 
-            self._links_gdf = GeoDataFrame(links_data, crs=self.lines_gdf.crs)
         return self._graph
 
+    def add_link(
+        self,
```

Comment:
If possible all these methods get a docstring (a one liner may do) and argument and return type annotations. That will make the code easier to follow for people that didn't write it.